aten/src/ATen/native/native_functions.yaml

```diff
@@ -4346,6 +4346,11 @@
   dispatch:
     SparseCPU, SparseCUDA: copy_sparse_
 
+- func: copy_opaque_to_opaque_(Tensor(a!) self, Tensor src, bool non_blocking=False) -> Tensor(a!)
+  variants: function
+  dispatch:
+    MkldnnCPU: copy_mkldnn_
```

Comment:
Don't do this. Instead, just add `MkldnnCPU: copy_mkldnn_` as an entry to `func: copy_` directly. Then you don't even need an if statement in the generic copy_ implementation. Your mkldnn kernel will then need to check that both inputs are mkldnn, due to multiple dispatch.

Reply:
Thanks for your comment!

>  Your mkldnn kernel will then need to check that both inputs are mkldnn, due to multiple dispatch.

Thanks, certainly it needs to be fixed.
I'd like to avoid multiple dispatch in one of the following ways:

(1) directly call `copy_mkldnn_` from `copy_impl` like [here](https://github.com/pytorch/pytorch/compare/master...msy-kato:PR_copy-mkldnn-tensor_try1).
This way is used for `quantized tensor` and `vulkan`.

(2) add `MkldnnCPU` to `copy_` in `native_functions.yaml` like [here](https://github.com/pytorch/pytorch/compare/master...msy-kato:PR_copy-mkldnn-tensor_try2) (you suggested).

I think (1) is a little better because, in the case (2), even if `self` is `MkldnnCPU` and `src` is normal Tensor,
`copy_mkldnn_` is dispatched and this is not the expected behavior.

```python
import torch
x = torch.randn(4, 5, dtype=torch.float32)
mkldnn_y = torch.randn(4, 5, dtype=torch.float32).to_mkldnn()
mkldnn_y.copy_(x) # call copy_mkldnn_ although both self and src are not mkldnn tensor 
```

gdb backtrace:

```
(gdb) bt
#0  0x00007fffeb416870 in at::native::copy_mkldnn_(at::Tensor&, at::Tensor const&, bool)@plt () from /home/users/pytorch/torch/lib/libtorch_cpu.so
#1  0x00007fffebf59555 in at::redispatch::copy_(c10::DispatchKeySet, at::Tensor&, at::Tensor const&, bool) () from /home/users/pytorch/torch/lib/libtorch_cpu.so
#2  0x00007fffee1bdd74 in torch::autograd::VariableType::(anonymous namespace)::copy_(c10::DispatchKeySet, at::Tensor&, at::Tensor const&, bool) () from /home/users/pytorch/torch/lib/libtorch_cpu.so
#3  0x00007fffec3d7a02 in at::Tensor::copy_(at::Tensor const&, bool) const () from /home/users/pytorch/torch/lib/libtorch_cpu.so
#4  0x00007ffff3da7f09 in torch::autograd::THPVariable_copy_(_object*, _object*, _object*) () from /home/users/pytorch/torch/lib/libtorch_python.so
```

I'd like to fix it (1) way, how about it?

Reply:
Well, (1) is certainly better than what you have here.

But I still prefer (2). Here's why. The if statement check to make sure that both arguments are MKLDNN is cheap, and the multiple dispatch gives you the option to implement CPU<->MKLDNN layout copies (if you support them). And the real benefit to doing it this way is it reduces the number of if statement checks we have to do in the base case (which @ngimel is always complaining about).

Reply:
Thanks for your advice.

> And the real benefit to doing it this way is it reduces the number of if statement checks we have to do in the base case

I understand. I applied the fix of (2).